### PR TITLE
X509: Fix handling of AKID and SKID extensions according to configuration

### PR DESCRIPTION
**Pull request by DDvO**
This (~fix for #16300~ **update:** simplified fix was extracted to #16442) was more work than expected because it turned out that the `req` and `x509` apps behaved on the AKID and SKID generation for self-signed certs in a way that was inconsistent with each other and with the documentation.
Now it should be as discussed in #16300.

This also fixes a bug in `v2i_AUTHORITY_KEYID()` on the use of `ASN1_STRING_length(ikeyid)` to check if the configuration indicatives "`none`" for the AKID extension.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by t-j-h** [comment]
Note - applied OTC decision label so this does not get approved and merged without explicit OTC approval.

File: doc/man5/x509v3_config.pod
Comment (vdukhovni) [inline doc/man5/x509v3_config.pod:202]:
```suggestion
(which is the same as the current certificate if it is self-signed), where
```

**Review by vdukhovni — CHANGES_REQUESTED** [review]
Looks good at first blush, just a minor nit I believe in the documentation.  I'd like to test this by building the pull request branch and seeing how it behaves now...

**Comment by vdukhovni** [comment]
A quick test shows slightly suprising, though improved behaviour.

With just `AKID = keyid` and no `SKID = hash` the self-signed cert still gets a SKID, but no AKID.  Otherwise, everything is as expected.

I would have expected the conditional AKID to happen, given that the SKID got generated anyway, or else for the SKID to also be skipped (since the AKID was conditional, and no explicit SKID was requested).

Of course the unexpected, but valid SKID is much better than the previous zero AKID, so definitely progress...

File: doc/man5/x509v3_config.pod
Comment (DDvO) [reply doc/man5/x509v3_config.pod:202]: I deliberately used "self-issued" at this point because it is more correct/general:
The point here is issuance (which means who signed the certificate).
Note that the cert may still not be self-signed in this case because a different pubkey may be placed in it, using `-force_pubkey`.

**Comment by DDvO** [comment]
> A quick test shows slightly suprising, though improved behaviour.

Thanks for trying it out and your feedback.

> With just `AKID = keyid` and no `SKID = hash` the self-signed cert still gets a SKID, but no AKID. Otherwise, everything is as expected.
> 
> I would have expected the conditional AKID to happen, given that the SKID got generated anyway, or else for the SKID to also be skipped (since the AKID was conditional, and no explicit SKID was requested).

Also this behavior is on purpose.

Since #13658, X509v3 certs obtain by default the key identifiers needed for RFC 5280 compliance.
For self-signed certs, giving the SKID only is sufficient and actually better because the AKID would be identical.

As documented,
* if you want to force inclusion of an AKID, use `authorityKeyIdentifier  = keyid:always`
* if you want to suppress the SKID (or AKID), use, e.g, `subjectKeyIdentifier = none`
Trying these out myself, I found a further minor bug for the latter case; a fixup is now added.

> Of course the unexpected,

It should not really be unexpected, as explained above.

> but valid SKID is much better than the previous zero AKID, so definitely progress...

This for sure, and 
the PR also fixes a couple of more bugs related to AKID and SKID generation for self-signed certs.

File: apps/req.c
Comment (levitte) [inline apps/req.c:845]: Uhm, doesn't this *take away* the possibility of a self-issued but not self-signed cert?

File: doc/man5/x509v3_config.pod
Comment (levitte) [reply doc/man5/x509v3_config.pod:202]: ... except that *in code*, you check with `cert_matches_key()` everywhere, which is effectively a check to see if the cert is self-*signed*, not self-issued.

Or am I missing something?

File: apps/req.c
Comment (DDvO) [reply apps/req.c:845]: No, it just precludes the generation of a (wrong) fallback AKID value for such certs.
See `v3_akid.c`:
```
        if (keyid == 2 && ikeyid == NULL && ctx->issuer_pkey != NULL) {
            /* generate fallback AKID, emulating s2i_skey_id(..., "hash") */
```
The `"Warning: Signature key and public key of cert do not match\n"` is not an error.

Comment (levitte) [reply apps/req.c:845]: Aha ok

File: doc/man5/x509v3_config.pod
Comment (DDvO) [reply doc/man5/x509v3_config.pod:202]: Things are a bit tricky, and it turns out that I had to further adapt both the doc and the code.
The check using `cert_matches_key()` prevents inclusion of a wrong fallback AKID for self-signed certs - this is fine here, but it turns out that for default AKID production `X509V3_set_issuer_pkey()` sill need to be called later.
For certs that are self-issued but not self-signed, the issuer cert may be given as a different cert with the same subject using the `-CA` option. This was so far not covered by the above doc line; I changed it to:
```
(which may be the same as the current certificate if it is self-issued), where
```
For the case explicitly mentioned there that we wanna produce an self-issued cert and the issuer cert is the same as the cert to be produced, the self-signed case was coded fine,
but when using `-force_pubkey` the SKID produced by default is wrong.
I plan to fix this further bug later today...

**Comment by mattcaswell** [comment]
Based on the OTC discussion yesterday I don't think this will be merged for 3.0.

File: doc/man5/x509v3_config.pod
Comment (vdukhovni) [reply doc/man5/x509v3_config.pod:202]: > ... except that _in code_, you check with `cert_matches_key()` everywhere, which is effectively a check to see if the cert is self-_signed_, not self-issued.
> 
> Or am I missing something?

You're not missing anything.  The documentation in question is trying to distinguish between certificates where the issuer key and name are in that same cert (it is self-signed) and other certificates (possibly self-issued) where the issuer key comes from a separate issuer certificate (which might share the same DN, but has a different key).  See RFC5280 Section 3.2:
```
   Self-signed certificates are self-issued certificates where the digital
   signature may be verified by the public key bound into the certificate.
```

Comment (vdukhovni) [reply doc/man5/x509v3_config.pod:202]: > I deliberately chose "self-issued" at this point because it is more correct/general:
> The point here is issuance (which means who signed the certificate).
> Note that a self-issued cert may still not be self-signed in this case because a different pubkey may be placed in it, e.g., using `-force_pubkey`.

That choice is wrong.  It is not the appropriate distinction.

**Comment by vdukhovni** [comment]
> Based on the OTC discussion yesterday I don't think this will be merged for 3.0.

Understood.  This does mean that some mildly misguided signer configurations will fail for self-signed certificates, that were "working" (producing a 0 AKID) with 1.1.1. But perhaps the view is that you don't want to be that much "bug compatible".

I do think that this change can go into a later patch release, it is I think a bugfix rather than a feature change.

File: doc/man5/x509v3_config.pod
Comment (DDvO) [reply doc/man5/x509v3_config.pod:202]: > > I deliberately chose "self-issued" at this point because it is more correct/general:
> > The point here is issuance (which means who signed the certificate).
> > Note that a self-issued cert may still not be self-signed in this case because a different pubkey may be placed in it, e.g., using `-force_pubkey`.
> 
> That choice is wrong. It is not the appropriate distinction.

Why do you think so?
How else would you address the case that a self-issued but not self-signed cert is produced?

Comment (vdukhovni) [reply doc/man5/x509v3_config.pod:202]: Self-issued certificates have an issuer CA cert that happens to have the same subject DN, but a different key and SKID.  AKID `keyid` handling is the same for cross certs and self-issued certs, the AKID keyid is a copy of the issuer's SKID.  The special case is self-signed, not self-issued.

**Comment by vdukhovni** [comment]
FWIW, I am fine with setting the "unexpected" SKID intentionally.

File: doc/man5/x509v3_config.pod
Comment (DDvO) [reply doc/man5/x509v3_config.pod:202]: > Self-issued certificates have an issuer CA cert that happens to have the same subject DN, but a different key and SKID.

Well, not quite: also self-signed certs (having same key and SKID) are self-issued,
while my question was how to cover in the above documentation sentence the case
of those self-issued certs that are not self-signed.

> AKID `keyid` handling is the same for cross certs and self-issued certs, the AKID keyid is a copy of the issuer's SKID. The special case is self-signed, not self-issued.

Good point that conceptually the special case for AKID production are self-signed certs,
simply because their AKID is typically omitted since it is the same as the SKID.

**Comment by DDvO** [comment]
Phew, some cases ware pretty hard to overlook and get right, in particular
the case that the issuer cert is the same as the subject cert and is self-issued but not self-signed.
Therefore I'm going to add some regression test cases.

**Comment by DDvO** [comment]
> Based on the OTC discussion yesterday I don't think this will be merged for 3.0.

This PR ends up fixing several in part pretty subtle bugs.
Which may be seen both as an argument for or against including it in 3.0.

Having a closer look, I also found that `X509V3_set_ctx()`, its use, and its documentation still was quite a mess,
which I believe is now cleaned up as well in this PR, in a separate commit.

File: doc/man5/x509v3_config.pod
Comment (DDvO) [reply doc/man5/x509v3_config.pod]: > AKID `keyid` handling is the same for cross certs and self-issued certs, the AKID keyid is a copy of the issuer's SKID. The special case is self-signed, not self-issued.

It turns out there is a further special case:
the case that the issuer cert is the same as the subject cert and is self-issued but not self-signed.
This is now fixed both in the code and the doc.

**Comment by DDvO** [comment]
@beldmit could you please make the external tests less brittle.
Checking for the presence of app warnings like
`"Warning: Signature key and public key of cert do not match\n"`
is likely a mistake.
This currently make the CI of this PR fail.

**Comment by beldmit** [comment]
AFAIR, this warning happens when the keys are not keypair. If you found a false positive. please let me know.

**Comment by vdukhovni** [comment]
> AFAIR, this warning happens when the keys are not keypair. If you found a false positive. please let me know.

David's note is pointing out that checking for expected errors SHOULD NOT look for a particular textual error message.  The error messages will vary as the code evolves.  Testing for expected errors should use a machine-readable result.

**Comment by DDvO** [comment]
> > AFAIR, this warning happens when the keys are not keypair. If you found a false positive. please let me know.
> 
> David's note is pointing out that checking for expected errors SHOULD NOT look for a particular textual error message. The error messages will vary as the code evolves. Testing for expected errors should use a machine-readable result.

Actually my point was about relying on warnings rather than errors,
but right that textual messages are particularly brittle.

The first of those failures of `test_external_gost_engine` is reported as follows:
```
=== Output failures of /home/runner/work/openssl/openssl/test-runs/test_external_gost_engine/tcl_tests/localhost-gost/ca.log ===
	???? 7: Displaying CRL start
	OPENSSL_CONF=/home/runner/work/openssl/openssl/gost-engine/tcl_tests/openssl-gost.cnf
	CRYPT_PARAMS is not set
	RNG is not set
	RNG_PARAMS is not set
	CCENGINE_LICENSE is not set
	OpenSSL cmdline: /home/runner/work/openssl/openssl/apps/openssl crl -text -noout -in test.crl
	???? 7: Displaying CRL ends failed
	Code:----
		grep "(Serial|Version|Signature Alg|Issuer)" [openssl "crl -text -noout -in test.crl"]
	---------------
	Expected status 0 got 0
	Expected result: {        Version 2 (0x1)
	        Signature Algorithm: GOST R 34.11-94 with GOST R 34.10-2001
	        Issuer: C = RU, L = Moscow, CN = Test CA gost2001:A, O = Cryptocom, OU = OpenSSL CA, emailAddress = openssl@cryptocom.ru
	    Serial Number: 011E
	STDERR CONTENTS:
	Warning: Signature key and public key of cert do not match    Signature Algorithm: GOST R 34.11-94 with GOST R 34.10-2001
	}
	     Got result: {        Version 2 (0x1)
	        Signature Algorithm: GOST R 34.11-94 with GOST R 34.10-2001
	        Issuer: C = RU, L = Moscow, CN = Test CA gost2001:A, O = Cryptocom, OU = OpenSSL CA, emailAddress = openssl@cryptocom.ru
	    Serial Number: 011E
	    Signature Algorithm: GOST R 34.11-94 with GOST R 34.10-2001
	}
```
and I find this pretty strange, for two reasons:
* the output is allegedly from the `crl` app, which is not touched by this PR
* the output suggests that the `"Warning: Signature key and public key of cert do not match\n"`was removed,
but in fact this PR does not remove it anywhere, but adds it (for self-issued certs) to the  `x509` app.

Apparently the output is misleading, and maybe the test takes the (changed) output of `x509` as reference for comparting with the (unchanged) output of `crl` or `ca`?
@beldmit, please provide clarification and an informed suggestion how to solve this.

**Comment by beldmit** [comment]
Thanks, I'll try to investigate what happens during the weekend, it looks pretty weird.

**Comment by DDvO** [comment]
I just fixed a mistake I introduced yesterday to `X509V3_set_ctx()`
and a bug in the `test_req` case `self-signed_v3_CA_both_KIDs.pem`.

**Comment by DDvO** [comment]
> Thanks, I'll try to investigate what happens during the weekend, it looks pretty weird.

Thanks @beldmit. I just noticed that the external-test failures look different to what I wrote above.

**Comment by DDvO** [comment]
Now
```
40871BA8D27F0000:error:1100007B:X509 V3 routines:v2i_AUTHORITY_KEYID:unable to get issuer keyid:crypto/x509/v3_akid.c:163:
/home/runner/work/openssl/openssl/gost-engine/test_tls.c:252: OpenSSL internal error: ext = X509V3_EXT_conf_nid(NULL, &v3ctx, NID_authority_key_identifier, "keyid:always,issuer")
```
is failing, maybe because "keyid:always" should be changed to "keyid" after the improved semantics of AKID generation?

**Comment by beldmit** [comment]
The failure looks relevant.

**Comment by beldmit** [comment]
@DDvO, does it mean that `keyid:always,issuer` is not acceptable now? Isn't it a regression, BTW?

And BTW, I don't see the TCL tests running at all so we still can have some more errors...

**Comment by DDvO** [comment]
> @DDvO, does it mean that `keyid:always,issuer` is not acceptable now? Isn't it a regression, BTW?

Without knowing what the test case exactly is, I cannot judge is the new code or the existing test.
Can you explain why you want/need to force the generation of the AKID, and where its value should come from? Is the cert self-signed,  or signed using a key belonging to some other issuing cert?

> And BTW, I don't see the TCL tests running at all so we still can have some more errors...

How to run/enable those tests,, and can you do?

**Comment by beldmit** [comment]
> > @DDvO, does it mean that `keyid:always,issuer` is not acceptable now? Isn't it a regression, BTW?
> 
> Without knowing what the test case exactly is, I cannot judge if the new code or the existing test is wrong.
> Can you explain why you want/need to force the generation of the AKID, and where its value should come from? Is the cert self-signed, or signed using a key belonging to some other issuing cert?

It should be a simple self-signed cert (and key)
https://github.com/gost-engine/engine/blob/master/test_tls.c#L252

**Comment by beldmit** [comment]
> > And BTW, I don't see the TCL tests running at all so we still can have some more errors...
> 
> How to run/enable those tests,, and can you do?

They will run if the c/perl tests succeed. The original issue (https://github.com/openssl/openssl/pull/16342#issuecomment-902489434) was generated by the TCL tests.

**Comment by DDvO** [comment]
> > Without knowing what the test case exactly is, I cannot judge if the new code or the existing test is wrong.
> > Can you explain why you want/need to force the generation of the AKID, and where its value should come from? Is the cert self-signed, or signed using a key belonging to some other issuing cert?
> 
> It should be a simple self-signed cert (and key)
> https://github.com/gost-engine/engine/blob/master/test_tls.c#L252

Thanks for this info. Thus I found that the problem is that in case issuer and subject cert are identical the code assumed that `X509V3_set_issuer_pkey()` has been called in order to be able to clearly check for self-signedness.
I've just relaxed this assumption, such that this issue should be gone now.

**Comment by beldmit** [comment]
Now we've got the same crl diagnostics as we met before :(

**Comment by DDvO** [comment]
> Now we've got the same crl diagnostics as we met before :(

So could you please analyzed this strange behavior and make sure that your test output can be understood better,
rather than puzzling readers as I explained above.

**Comment by beldmit** [comment]
I'll try.

**Comment by beldmit** [comment]
Sure. I'm sorry, the tests have rather ancient history and full of "here be dragons" code (and logic) snippets.

Here is the failing test: 
https://github.com/gost-engine/engine/blob/879e6111068536f27686223602cd1e4789331632/tcl_tests/ca.try#L75-L81

The problematic part of the report is present in the variable $serial_list

It is filled before, see

https://github.com/gost-engine/engine/blob/879e6111068536f27686223602cd1e4789331632/tcl_tests/ca.try#L61-L67

So the problematic command that started generating the warning is with high probability
`openssl "x509 -in $revoking_cert -noout -serial"`

Could you please check whether we should have this warning here? I don't think so, because I don't see where we could get a private key for comparison.

**Comment by DDvO** [comment]
> So the problematic command that started generating the warning is with high probability
> `openssl "x509 -in $revoking_cert -noout -serial"`

I see!
Would be good if your test output contained a hint that the expected output is produced this way, using `x509`.
This is crucial for understanding what is going on here, in particular why additional output of `x509` can make these sort of tests fail.

> Could you please check whether we should have this warning here? I don't think so, because I don't see where we could get a private key for comparison.

Indeed it makes no sense to give this warning in cases where `x509` is not provided with a private key.
I've extended the guard of the warning output accordingly, so this problem should be gone now.

File: apps/x509.c
Comment (beldmit) [inline apps/x509.c:813]: Could we move this check inside the cert_matches_key?

Comment (DDvO) [reply apps/x509.c:813]: This check is already (indirectly) in there, making the function return 0 if privkey equals NULL.
Yet here we need to suppress the warning in this case.

Comment (beldmit) [reply apps/x509.c:813]: `cert_matches_key` looks like the apps-purposed wrapper around `X509_check_private_key` with the only purpose - raise a warning on mismatch. As we shouldn't raise this warning in absence of a private key, and, if we should get an error, we could call `X509_check_private_key` directly, i'd suggest returning 0 from `cert_matches_key` when privkey is not available.

File: apps/ca.c
Comment (beldmit) [inline apps/ca.c:1245]: Add a check here?

File: apps/cmp.c
Comment (beldmit) [inline apps/cmp.c:1588]: And here...

File: apps/pkcs12.c
Comment (beldmit) [inline apps/pkcs12.c:558]: `cert_matches_key` will clear the error raised by X509_check_private_key, that may make sense...

File: apps/req.c
Comment (beldmit) [inline apps/req.c:885]: Check result?

File: crypto/x509/v3_akid.c
Comment (beldmit) [inline crypto/x509/v3_akid.c:136]: And this block looks exactly like `cert_matches_key` to me

**Review by beldmit — CHANGES_REQUESTED** [review]

File: apps/ca.c
Comment (vdukhovni) [inline apps/ca.c:1713]: It is not entirely clear to me why the `req` argument was replaced with NULL.  Is it redundant here?  Why?

File: apps/x509.c
Comment (DDvO) [reply apps/x509.c:813]: > `cert_matches_key` looks like the apps-purposed wrapper around `X509_check_private_key` with the only purpose - raise a warning on mismatch.

Not really - `cert_matches_key()` is a wrapper just for preventing spurious error queue entries.
Note that it is going to be used also in `apps/pkcs12.c` in a situation where a warning would not be appropriate.

> As we shouldn't raise this warning in absence of a private key, and, if we should get an error, we could call `X509_check_private_key` directly, i'd suggest returning 0 from `cert_matches_key` when privkey is not available.

As mentioned, `X509_check_private_key()` and thus `cert_matches_key()` already returns 0 in case the key parameter is NULL.
And this does not relieve us from adding the guard `privkey != NULL`, as is done now, 
since otherwise we would get an unsuitable warning which also confused your tests.

Comment (DDvO) [reply apps/x509.c:813]: BTW, @beldmit, did you add to, e.g., https://github.com/gost-engine/engine/blob/879e6111068536f27686223602cd1e4789331632/tcl_tests/ca.try#L61-L67
a hint that the expected output for these test cases is derived using, e..g, 
`openssl x509 -in $revoking_cert -noout -serial"`?

Comment (beldmit) [reply apps/x509.c:813]: You are right, will do.

File: apps/ca.c
Comment (DDvO) [reply apps/ca.c]: All right, added the possibility for `X509V3_set_nconf()` to return error.
Also simplified its use in `x509.c`, `req.c`, and `ca.c`.

File: apps/cmp.c
Comment (DDvO) [reply apps/cmp.c]: Done.

File: apps/pkcs12.c
Comment (DDvO) [reply apps/pkcs12.c:558]: Yes, that was the purpose of this change.

File: apps/req.c
Comment (DDvO) [reply apps/req.c]: Done

File: crypto/x509/v3_akid.c
Comment (DDvO) [reply crypto/x509/v3_akid.c:136]: Yes, but `cert_matches_key()` is defined just for apps (in `apps/lib/apps.c`).

Comment (beldmit) [reply crypto/x509/v3_akid.c:136]: Ah! Sorry, my fault

File: apps/ca.c
Comment (DDvO) [reply apps/ca.c:1705]: I removed it because otherwise `s2i_skey_id()` in `crypto/x509/v3_skid.c` may pick up the wrong value - 
here the one in `x509` is relevant.
Note that this PR now enforces and documents that at most one of the certificate (`subject`), the certification request (`req`), or the certificate revocation list (`crl`) parameters can be non-NULL.

Comment (vdukhovni) [reply apps/ca.c:1705]: It'd be good to add a  brief comment that noted why the `req` argument is NULL, to avoid future regressions.

Comment (DDvO) [reply apps/ca.c:1705]: One could do this, but meanwhile `X509V3_set_ctx()` would throw an error if `req` was given in addition to a reference cert or CRL.

**Comment by DDvO** [comment]
> Phew, some cases ware pretty hard to overlook and get right, in particular
> the case that the issuer cert is the same as the subject cert and is self-issued but not self-signed.
> Therefore I'm going to add some regression test cases.

I just added systematic tests for AKID+SKID generation for self-signed certs.
Some more to come for self-issued(-only) certs and non-self-issued certs....

**Comment by mattcaswell** [comment]
OTC: Moving this to post 3.0 milestone

**Comment by DDvO** [comment]
Removing "triaged: bug" is not right - 
this PR fixes multiple obscure bugs in SKID/AKID generation".

**Comment by t8m** [comment]
It is not that important if this mixes bug fixes with the improvement feature. The problem is we cannot merge this to 3.0 because it contains the feature change. The additional triaged: bug label would not add much value anyway.

**Comment by DDvO** [comment]
Rebased this PR, and in order to simplify it, 
extracted two commits (with refactoring and improved error handling) to #16998.
Ping for continuation of the review.

File: apps/ca.c
Comment (vdukhovni) [inline apps/ca.c:1711]: Perhaps add a comment that all the necessary information is in `ret`, so there's no need to also provide `req` (assuming this is why `req` was replaced with NULL in the arg list).

**Review by vdukhovni — APPROVED** [review]
The C code changes look fine.  I've not looked at the test changes in detail. Hope that's OK.

File: apps/ca.c
Comment (DDvO) [reply apps/ca.c:1711]: Good point - done.

**Comment by DDvO** [comment]
> The C code changes look fine.


Very pleased to hear!

**Review by t8m — APPROVED** [review]
OK for master.

**Review by beldmit — APPROVED** [review]
LGTM, though I don't understand this stuff deeply.

**Comment by DDvO** [comment]
What does the label **hold: merge after β period** currently mean?
Is 3.1-dev (the master branch) the current beta?

**Comment by t8m** [comment]
> What does the label **hold: merge after β period** currently mean?

That label is obsolete now. I am going to remove it.

**Comment by t8m** [comment]
In the end I did not remove the label (just slightly adjusted its name) as it can be helpful in future. I've just removed it from the existing pull requests.

**Comment by DDvO** [comment]
Merged to master.
Thanks for the discussion and reviews.

**Pull request closed without merging**
